pkg/api/store/nodetemplate/store.go

```diff
@@ -24,43 +24,95 @@ type nodeTemplateStore struct {
-	}
-	return &transform.Store{
-		Store:       s,
-		Transformer: s.filterLegacyTemplates,
 	}
 }
 
-func (s *nodeTemplateStore) filterLegacyTemplates(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}, opt *types.QueryOptions) (map[string]interface{}, error) {
+func (t *transformer) filterLegacyTemplates(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}, opt *types.QueryOptions) (map[string]interface{}, error) {
 	ns, _ := data["namespaceId"].(string)
-	name, _ := data["name"].(string)
+	id, _ := data["id"].(string)
 	if ns != namespace.NodeTemplateGlobalNamespace {
-		s.NodeTemplateClient.Controller().Enqueue(ns, name)
+		// nodetemplate may not have been migrated yet, enqueue
+		t.NodeTemplateClient.Controller().Enqueue(ns, strings.TrimPrefix(id, ns+":"))
 		return nil, nil
 	}
 	return data, nil
 }
 
+func (s *nodeTemplateStore) ByID(apiContext *types.APIContext, schema *types.Schema, id string) (map[string]interface{}, error) {
+	originalNs, originalID := getNsAndID(id)
+	fullMigratedID := getMigratedID(originalNs, originalID)
+	migratedNs, migratedID := getNsAndID(fullMigratedID)
+	data, err := s.Store.ByID(apiContext, schema, fullMigratedID)
+	if err != nil {
+		return nil, replaceIDInError(err, migratedID, originalID, migratedNs, originalNs)
+	}
+
+	data["namespaceId"] = originalNs
+	data["id"] = id
+	return data, nil
+}
+
+func getMigratedID(ns, id string) string {
+	if ns == namespace.NodeTemplateGlobalNamespace {
+		return fmt.Sprintf("%s:%s", ns, id)
+	}
+
+	return fmt.Sprintf("%s:nt-%s-%s", namespace.NodeTemplateGlobalNamespace, ns, id)
+}
+
+func getNsAndID(id string) (string, string) {
+	idParts := strings.SplitN(id, ":", 2)
+	if len(idParts) == 1 {
+		return "", idParts[0]
+	}
+	return idParts[0], idParts[1]
+}
+
+func replaceIDInError(err error, id, newID, ns, newNs string) error {
+	if apiError, ok := err.(*httperror.APIError); ok {
+		msg := strings.Replace(apiError.Message, id, newID, -1)
+		msg = strings.Replace(msg, ns, newNs, -1)
+		err = httperror.NewAPIError(apiError.Code, msg)
+	}
+	return err
+}
+
 func (s *nodeTemplateStore) Delete(apiContext *types.APIContext, schema *types.Schema, id string) (map[string]interface{}, error) {
+	originalNs, originalID := getNsAndID(id)
+	fullMigratedID := getMigratedID(originalNs, originalID)
+	migratedNs, migratedID := getNsAndID(fullMigratedID)
 	pools, err := s.NodePoolLister.List("", labels.Everything())
 	if err != nil {
 		return nil, err
 	}
 	for _, pool := range pools {
-		if pool.Spec.NodeTemplateName == id {
+		if pool.Spec.NodeTemplateName == fullMigratedID {
 			return nil, httperror.NewAPIError(httperror.MethodNotAllowed, "Template is in use by a node pool.")
 		}
 	}
-	return s.Store.Delete(apiContext, schema, id)
+	data, err := s.Store.Delete(apiContext, schema, fullMigratedID)
+	if err != nil {
+		return nil, replaceIDInError(err, migratedID, originalID, migratedNs, originalNs)
+	}
+	return data, nil
```

Comment:
replace back the originalId and NS in the response data?

Reply:
yes, id is being replaced to not break anything that may rely or reuse response id- as request by @ibuildthecloud and @cloudnautique

Reply:
right understood, but I meant why are we not replacing it back here for Delete() like its done in Update()

Reply:
oh- because data return by deletes is nil anyway. I can just return a nil to be safe